src/main/java/agh/project/Elements/Animal.java

```diff
@@ -0,0 +1,182 @@
+    private int energy;
+    private Genes genes;
+    private int grassEaten = 0;
+    private int kids = 0;
+    private int age = 0;
+
+    private ArrayList<IPositionChangeObserver> observers = new ArrayList<>();
+
+    public Animal(IWorldMap map, Vector2d position, int energy) {
+        this.map = map;
+        this.position = position;
+        this.energy = energy;
+        this.genes = new Genes();
+        addObserver(this.map);
+    }
+
+    public int getActiveGenIndex() { // tu trzeba się zastanowić
+        return this.genes.getActiveGenIndex();
+    }
+
+    //move function
+    public void move() {
+        List<Integer> genesList = genes.getGenes();
+        int genIndex = this.genes.getActiveGenIndex();
+        int gen = genesList.get(genIndex);
+
+        //temporary position
+        Vector2d tempPosition = this.position;
+
+        //old position
+        Vector2d oldPosition = this.position;
+
+        switch (gen) {
+            case 0 -> position.add(this.direction.toUnitVector());
+            case 1 -> {
+                direction.next();
+                tempPosition = tempPosition.add(this.direction.toUnitVector());
+
+            }
+            case 2 -> {
+                direction.next().next();
+                tempPosition = tempPosition.add(this.direction.toUnitVector());
+            }
+            case 3 -> {
+                direction.next().next().next();
+                tempPosition = tempPosition.add(this.direction.toUnitVector());
+            }
+            case 4 -> {
+                tempPosition = tempPosition.subtract(this.direction.toUnitVector());
+            }
+            case 5 -> {
+                direction.previous().previous().previous();
+                tempPosition = tempPosition.add(this.direction.toUnitVector());
+            }
+            case 6 -> {
+                direction.previous().previous();
+                tempPosition = tempPosition.add(this.direction.toUnitVector());
+            }
+            case 7 -> {
+                direction.previous();
+                tempPosition = tempPosition.add(this.direction.toUnitVector());
+            }
+        }
+
+        //jeżeli nie może się ruczyć w dane miejsce(chce wyjść poza mapę)
+        if (!this.map.canMoveTo(this.position)) {
+
+            //wtedy użyj spejalnej metody poruszania (w zależności od wariantu mapy)
+            this.map.specialMoves(this, tempPosition);
+
+        } else { // wprzeciwnym przypadku zwierzę rusza się w obrębie mapy więc ruch jest poprawny
+            this.position = tempPosition;
+        }
+
+        int activeGenIndex = this.getActiveGenIndex();
+        this.genes.nextGen(activeGenIndex);
+        this.ageIncrement();
+        this.removeEnergy(moveEnergy);
+        positionChanged(oldPosition);
+    }
+
+    public int getAge() {
+        return age;
+    }
+
+    @Override
+    public int compareTo(Animal other) {
+        return this.energy - other.energy;
```

Comment:
czemu akurat po energii? z tego co widzę to w mapie macie logikę sortowania, która trochę korzysta z tego, a trochę i tak dorzuca komparatory -> można by to w całości zrobić na komparatorach, chyba że domyślne układanie po energii gdzieś jeszcze się przydaje.